pkg/volume/empty_dir/quota.go

```diff
@@ -0,0 +1,162 @@
+	if medium == api.StorageMediumMemory {
+		glog.V(3).Infof("skipping quota application due to memory storage medium.")
+		return nil
+	}
+	isXFS, err := xqa.isXFS(dir)
+	if err != nil {
+		return err
+	}
+	if !isXFS {
+		return fmt.Errorf("unable to apply quota: %s is not on an XFS filesystem", dir)
+	}
+	if fsGroup == nil {
+		// This indicates the operation matched an SCC with FSGroup strategy RunAsAny.
+		// Not an error condition.
+		glog.V(3).Infof("Unable to apply XFS quota, no FSGroup specified.")
+		return nil
+	}
+
+	volDevice, err := xqa.getFsDevice(dir)
+	if err != nil {
+		return err
+	}
+
+	err = xqa.applyQuota(volDevice, quota, *fsGroup)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Check if the empty dir is on an XFS filesystem:
+func (xqa *xfsQuotaApplicator) isXFS(dir string) (bool, error) {
+	out, _, err := xqa.cmdRunner.RunFsTypeCommand(dir)
+	if err != nil {
+		glog.Errorf("error checking filesystem type for %s: %s", dir, err)
+		return false, fmt.Errorf("unable to check filesystem type for emptydir volume %s: %s", dir, err)
+	}
+	if strings.Contains(out, "xfs") {
+		return true, nil
+	}
+	return false, nil
+}
+
+func (xqa *xfsQuotaApplicator) applyQuota(volDevice string, quotaInMB int, fsGroupID int64) error {
+	_, stderr, err := xqa.cmdRunner.RunApplyQuotaCommand(volDevice, quotaInMB, fsGroupID)
+	if err != nil {
+		glog.Errorf("Error running xfs_quota: %s", err)
+		return err
+	}
+	// xfs_quota is very happy to fail but return a success code, likely due to its
+	// interactive shell approach. Grab stderr, if we see anything written to it we'll
+	// consider this an error.
+	if len(stderr) > 0 {
+		glog.Error("Error running xfs_quota: ", stderr)
+		return errors.New("xfs_quota wrote to stderr")
+	}
+
+	glog.V(3).Infof("XFS quota applied: device=%s, quota=%dMB, fsGroup=%d", volDevice, quotaInMB, fsGroupID)
+	return nil
+}
+
+// Returns the fs device for a given emptydir path. To do this we
+// run df on the emptydir path, returning a header line and the line we're
+// interested in. The first string token in that line will be the device name.
+func (xqa *xfsQuotaApplicator) getFsDevice(dir string) (string, error) {
+	out, _, err := xqa.cmdRunner.RunFsDeviceCommand(dir)
+	if err != nil {
+		return "", fmt.Errorf("unable to find filesystem device for emptydir volume %s: %s", dir, err)
+	}
+	fsDevice, parseErr := parseFsDevice(out)
+	return fsDevice, parseErr
+}
+
+// Extracts the first lines device from df output:
+func parseFsDevice(dfOutput string) (string, error) {
+	// Need to skip the df header line starting with "Filesystem", and grab the first
+	// word of the following line which will be our device path.
+	lines := strings.Split(dfOutput, "\n")
+	if len(lines) < 2 {
+		glog.Errorf("Error finding volume device in df output: %s", dfOutput)
+		return "", errors.New("insufficient lines in df output")
+	}
+
+	fsDevice := strings.Split(lines[1], " ")[0]
```

Comment:
I think this might be correct as is, lines[1] should be safe via the check above, and splitting any string should always return a [0] as far as I can tell. Added some more tests to try to make sure but let me know if I missed something in this.